tools/codegen/gen.py

```diff
@@ -462,10 +462,15 @@ def __call__(self, f: NativeFunction) -> Optional[str]:
 
         assert self.target is Target.DEFINITION
 
-        def generate_defn(sig: CppSignature) -> str:
+        def generate_defn(faithful: bool) -> str:
             dispatcher_sig = DispatcherSignature.from_schema(f.func)
 
-            dispatcher_exprs = dispatcher.cpparguments_exprs(sig.argument_packs())
+            if faithful and sig_group.faithful_signature is not None:
+                sig = sig_group.faithful_signature
+            else:
+                sig = sig_group.signature
+
+            dispatcher_exprs = dispatcher.cpparguments_exprs(f.func, method=False, api_is_faithful=faithful)
```

Comment:
Let's go thru this one together - it's worth working back from here to the things it's calling and figuring out how faithfulness can be better encapsulated.

Reply:
as discussed offline, part of the complexity here comes from the fact that we need to match 2x2=4 scenarios. The API can be faithful/unfaithful and the dispatcher calling convention can also be faithful/unfaithful. This will get cleaner once the dispatcher calling convention is always faithful.